### PR TITLE
circt-{as,dis}: init basic MLIR <--> MLIRBC utilities.

### PR DESCRIPTION
**Pull request by dtzSiFive**
Basic utilities for converting to/from bytecode.

`circt-opt` can do the same, but these make it more explicit and are simpler to invoke.

### CODE REVIEW

**Comment by teqdruid** [comment]
Just out of curiosity, what happens if you don't register a dialect? Does it output the generic form?

**Comment by dtzSiFive** [comment]
> Just out of curiosity, what happens if you don't register a dialect? Does it output the generic form?

Good question!

As-is an error is generated complaining about the unknown op, in both utilities.

Poking a bit, calling `allowUnregisteredDialects()` on the `MLIRContext` in `circt-dis` causes it to print the generic form for the ops from unregistered dialects.  `circt-as` (writing to bytecode) seems to require the dialects to be registered in my limited testing.  I'm not sure what the explanation is, or if this is something to be relied upon-- for example if we implement the `BytecodeDialectInterface` to define custom encodings for our types/attributes, I wouldn't expect to be able to decode without having that interface and its decode/read hook available.

**Comment by teqdruid** [comment]
> > Just out of curiosity, what happens if you don't register a dialect? Does it output the generic form?
> 
> Good question!
> 
> As-is an error is generated complaining about the unknown op, in both utilities.
> 
> Poking a bit, calling `allowUnregisteredDialects()` on the `MLIRContext` in `circt-dis` causes it to print the generic form for the ops from unregistered dialects. `circt-as` (writing to bytecode) seems to require the dialects to be registered in my limited testing. I'm not sure what the explanation is, or if this is something to be relied upon-- for example if we implement the `BytecodeDialectInterface` to define custom encodings for our types/attributes, I wouldn't expect to be able to decode without having that interface and its decode/read hook available.

Yeah, I've been wondering what MLIR bytecode would do for Types/Attributes since they are specified exclusively in C++ and thus don't have a generic form like ops. Apparently, the answer is link in some custom code (which doesn't surprise me).

**Review by uenoku — APPROVED** [review]
Thank you for adding these handy tools! Can we add a simple test for it? 
I'm in flavor of having these tools as separated tool from circt-opt.

File: tools/circt-dis/circt-dis.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  module->print(output->os());
+  output->keep();
+
+  return success();
+}
+
+int main(int argc, char **argv) {
+  InitLLVM y(argc, argv);
+  mlir::DialectRegistry registry;
+
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply tools/circt-dis/circt-dis.cpp:100]: The first one (registering cl options for MLIRContext) adds options for disabling threading and print-{op,stacktrace}-on-diagnostic which didn't seem particularly useful/relevant for these simple tools, but don't feel strongly about omitting it.  AsmPrinter, especially for `circt-dis` sounds like a good idea!  Lets you specify printing generic form, for example.

File: test/circt-as-dis/basic-firrtl.mlir [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// RUN: circt-as %s -o - | circt-opt | FileCheck %s
```
Comment (uenoku) [inline test/circt-as-dis/basic-firrtl.mlir:1]: Nice!

**Pull request merged**